File: go/2022/day03/day03.go
package main

import (
	"flag"
	"fmt"
	"os"
	"strings"
)

var inputFile = flag.String("inputFile", "input.txt", "Relative file path to use as input")

func main() {
	flag.Parse()
	bytes, err := os.ReadFile(*inputFile)
	if err != nil {
		panic("Unable to read the input file")
	}

	input := string(bytes)

	fmt.Printf("AoC 2022, Day3, Part1 solution is: %d \n", part1(input))
	fmt.Printf("AoC 2022, Day3, Part2 solution is: %d \n", part2(input))
}

func part1(input string) int {
	priorities := getPriorities()
	total := 0

	for _, line := range strings.Split(input, "\n") {
		m := len(line) / 2
		commonChars := getCommonChar(line[:m], line[m:])
		if len(commonChars) != 1 {
			panic("expected a single char")
		}
		total += priorities[commonChars[0]]
	}

	return total

}

func part2(input string) int {
	priorities := getPriorities()
	total := 0

	lines := strings.Split(input, "\n")

	for i := 2; i < len(lines); i += 3 {
		str1 := lines[i-2]
		str2 := lines[i-1]
		str3 := lines[i]

		c1 := getCommonChar(str1, str2)
		c2 := getCommonChar(str3, str1)
		commonChars := getCommonChar(strings.Join(c1, ""), strings.Join(c2, ""))

		if len(commonChars) != 1 {
			panic("expected a single char")
		}
		total += priorities[commonChars[0]]
	}
	return total

}

func getCommonChar(str1, str2 string) []string {
	copart1Map := getStringSet(str1)
	commonChars := []string{}
	uniqueChars := map[rune]bool{}

	for _, c := range str2 {
		if copart1Map[c] {
			if !uniqueChars[c] {
				commonChars = append(commonChars, string(c))
				uniqueChars[c] = true
			}
		}
	}
	return commonChars
}

func getPriorities() map[string]int {
	priorities := map[string]int{}
	for i := 0; i < 26; i++ {
		priorities[string(rune('a'+i))] = i + 1
		priorities[string(rune('A'+i))] = i + 27
	}

	return priorities
}

func getStringSet(input string) map[rune]bool {
	occurrence := map[rune]bool{}

	for _, c := range input {
		occurrence[c] = true
	}
	return occurrence
}

File: python/2022/day03/day03.py
from typing import List
from string import ascii_lowercase, ascii_uppercase

key = ascii_lowercase + ascii_uppercase


def part1(input: str) -> int:
    total = 0
    for line in parse_input(input):
        n = len(line)

        sub1 = line[: (n // 2)]
        sub2 = line[(n // 2) :]

        c = list(set(sub1).intersection(sub2))
        total += key.index(c[0]) + 1

    return total


def part2(input: str) -> int:
    lines = parse_input(input)

    total = 0
    for i in range(0, len(lines), 3):
        group = lines[i : (i + 3)]
        c = list(set(group[0]).intersection(group[1]).intersection(group[2]))
        total += key.index(c[0]) + 1

    return total


def parse_input(input: str) -> List[str]:
    with open(input) as fn:
        raw_data = fn.read().strip()
        parts = raw_data.split("\n")
    return parts


def main():
    puzzle_input_path = '2022/day03/input.txt'
    print(f"AoC2022 Day3, Part1 solution is: {part1(puzzle_input_path)}")
    print(f"AoC2022 Day3, Part2 solution is: {part2(puzzle_input_path)}")


if __name__ == "__main__":
    main()
